server_cached_router.go

```diff
@@ -0,0 +1,230 @@
+func (r cachedRouter) FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[*types.PeerRecord], error) {
+	it, err := r.router.FindPeers(ctx, pid, limit)
+
+	if err == routing.ErrNotFound {
+		// ErrNotFound will be returned if either dialing the peer failed or the peer was not found
+		r.cachedAddrBook.RecordFailedConnection(pid) // record the failure used for probing/backoff purposes
+		return nil, routing.ErrNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// update the metrics to indicate that we didn't look up the cache for this lookup
+	peerAddrLookups.WithLabelValues(addrCacheStateUnused, addrQueryOriginPeers).Inc()
+	return it, nil
+}
+
+// withAddrsFromCache returns the best list of addrs for specified [peer.ID].
+// It will consult cache ONLY if the addrs slice passed to it is empty.
+func (r cachedRouter) withAddrsFromCache(queryOrigin string, pid peer.ID, addrs []types.Multiaddr) []types.Multiaddr {
+	// skip cache if we already have addrs
+	if len(addrs) > 0 {
+		peerAddrLookups.WithLabelValues(addrCacheStateUnused, queryOrigin).Inc()
+		return addrs
+	}
+
+	cachedAddrs := r.cachedAddrBook.GetCachedAddrs(pid)
+	if len(cachedAddrs) > 0 {
+		logger.Debugw("found cached addresses", "peer", pid, "cachedAddrs", cachedAddrs)
+		peerAddrLookups.WithLabelValues(addrCacheStateHit, queryOrigin).Inc()
+		return cachedAddrs
+	} else {
+		// Cache miss. Queue peer for lookup.
+		peerAddrLookups.WithLabelValues(addrCacheStateMiss, queryOrigin).Inc()
+		return nil
+	}
+}
+
+var _ iter.ResultIter[types.Record] = &cacheFallbackIter{}
+
+type cacheFallbackIter struct {
+	sourceIter      iter.ResultIter[types.Record]
+	current         iter.Result[types.Record]
+	findPeersResult chan types.PeerRecord
+	router          cachedRouter
+	ctx             context.Context
+	cancel          context.CancelFunc
+	ongoingLookups  atomic.Int32
+}
+
+// NewCacheFallbackIter is a wrapper around a results iterator that will resolve peers with no addresses from cache and if no cached addresses, will look them up via FindPeers.
+// It's a bit complex because it ensures we continue iterating without blocking on the FindPeers call.
+func NewCacheFallbackIter(sourceIter iter.ResultIter[types.Record], router cachedRouter, ctx context.Context) *cacheFallbackIter {
+	ctx, cancel := context.WithCancel(ctx)
+	return &cacheFallbackIter{
+		sourceIter:      sourceIter,
+		router:          router,
+		ctx:             ctx,
+		cancel:          cancel,
+		findPeersResult: make(chan types.PeerRecord),
+		ongoingLookups:  atomic.Int32{},
+	}
+}
+
+func (it *cacheFallbackIter) Next() bool {
+	// Try to get the next value from the source iterator first
+	if it.sourceIter.Next() {
+		val := it.sourceIter.Val()
+		handleRecord := func(id *peer.ID, record *types.PeerRecord) bool {
+			record.Addrs = it.router.withAddrsFromCache(addrQueryOriginProviders, *id, record.Addrs)
+			if record.Addrs != nil {
```

Comment:
nit: more idiomatic go is to avoid nil checks, length ones are less error prone during refactors (len will work with nil and empty slice)
```suggestion
			if len(record.Addrs) != 0 {
```